localstack/services/sns/publisher.py

```diff
@@ -0,0 +1,946 @@
+import abc
+import ast
+import base64
+import datetime
+import json
+import logging
+import time
+import traceback
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from typing import Dict, List, Tuple, Union
+
+import requests
+
+from localstack import config
+from localstack.aws.api.lambda_ import InvocationType
+from localstack.aws.api.sns import MessageAttributeMap
+from localstack.aws.api.sqs import MessageBodyAttributeMap
+from localstack.config import external_service_url
+from localstack.services.sns import constants as sns_constants
+from localstack.services.sns.models import (
+    SnsApplicationPlatforms,
+    SnsMessage,
+    SnsStore,
+    SnsSubscription,
+)
+from localstack.utils.aws import aws_stack
+from localstack.utils.aws.arns import (
+    extract_region_from_arn,
+    extract_resource_from_arn,
+    parse_arn,
+    sqs_queue_url_for_arn,
+)
+from localstack.utils.aws.aws_responses import create_sqs_system_attributes
+from localstack.utils.aws.dead_letter_queue import sns_error_to_dead_letter_queue
+from localstack.utils.cloudwatch.cloudwatch_util import store_cloudwatch_logs
+from localstack.utils.json import json_safe
+from localstack.utils.objects import not_none_or
+from localstack.utils.strings import long_uid, md5, to_bytes
+from localstack.utils.time import timestamp_millis
+
+LOG = logging.getLogger(__name__)
+
+# future config flag
+PLATFORM_APPLICATION_REAL = False
+
+
+@dataclass(frozen=True)
+class SnsPublishContext:
+    message: SnsMessage
+    store: SnsStore
+    request_headers: Dict[str, str]
+
+
+@dataclass
+class SnsBatchPublishContext:
+    messages: List[SnsMessage]
+    store: SnsStore
+    request_headers: Dict[str, str]
+
+
+class TopicPublisher(abc.ABC):
+    def publish(self, context: SnsPublishContext, subscriber: SnsSubscription):
+        try:
+            self._publish(context=context, subscriber=subscriber)
+        except Exception:
+            LOG.exception(
+                "An internal error occurred while trying to send the SNS message %s",
+                context.message,
+            )
+            return
```

Comment:
I'm not quite sure I understand fully the concept between caller/callee in this case.

The only way the caller could manage this would be by checking the `Future` result, or wrapping like you said.
I guess the `PublishDispatcher` could theoretically provide a wrapper method, something like `_wrap_and_log()` or something similar, and dispatch it that way? `executor.submit(self._wrap_and_log(publisher.publish))`, is that what you meant?
We could almost provide an helper method and use it as a decorator around the `publish` methods? I'll wait for your answer on this one as I'm not sure I fully understood what you meant! But thanks, agree that the logging is a bit weird in the publisher that way.

Reply:
Update: we discussed and we will tackle this in a next PR, this will do for now. Will add a comment and a `todo` to clean up this.